server/src/main/java/org/opensearch/tasks/TaskManager.java

```diff
@@ -448,6 +491,99 @@ public void waitForTaskCompletion(Task task, long untilInNanos) {
         throw new OpenSearchTimeoutException("Timed out waiting for completion of [{}]", task);
     }
 
+    /**
+     * Adds Task Id in the ThreadContext.
+     * <p>
+     * Stashes the existing ThreadContext and preserves all the existing ThreadContext's data in the new ThreadContext
+     * as well.
+     *
+     * @param task for which Task Id needs to be added in ThreadContext.
+     * @return StoredContext reference to restore the ThreadContext from which we created a new one.
+     * Caller can call context.restore() to get the existing ThreadContext back.
+     */
+    public ThreadContext.StoredContext addTaskIdInThreadContext(@Nullable Task task) {
+        if (task == null) {
+            return () -> {};
+        }
+
+        ThreadContext threadContext = threadPool.getThreadContext();
+
+        if (threadContext.getTransient(TASK_ID) != null) {
+            logger.warn(
+                "Task Id already present in the thread context. Thread Id: {}, Existing Task Id: {}, New Task Id: {}. Overwriting",
+                Thread.currentThread().getId(),
+                threadContext.getTransient(TASK_ID),
+                task.getId()
+            );
```

Comment:
Is this expected to be idempotent? If yes we can push for stronger assertions

Reply:
Still verifying all cases where this would happen by running this change on a multi node cluster.

Reply:
This was majorly for tasks which executes child task synchronously.